### PR TITLE
Fix recursive deep scan of buckets

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Deep scanning of buckets did not perform deep scan.

Pass deep scan parameter down to `listObjects` so it will not only return objects that failed quorum checks, but all objects.

## How to test this PR?

* Start server in XL mode.
* Upload a file.
* Delete a part of a file.
* Run `mc admin heal myminio/bucket -recursive -scan=deep`

Before: Missing part would not be detected.
After: Missing part is restored.

Of course this will put more load on the server, but `-scan=deep` requested a deep scan.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#9c590065-cd1f-4474-bfee-ba8888a4cc56) |
| mint-gateway-azure.sh | :x: [more...](#f95c5552-3b06-4341-b657-3f9f20dd3545) |
<a name="9c590065-cd1f-4474-bfee-ba8888a4cc56"></a>
### 8900-4fb7909/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:31081
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f7edca825893:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 24 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 8 seconds
(4/13) Running mc tests ... done in 57 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 30 seconds
{
  "args": {},
  "duration": 84684,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="f95c5552-3b06-4341-b657-3f9f20dd3545"></a>
### 8900-4fb7909/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:32686
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 89af43756a76:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 6 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 15 seconds
(4/13) Running mc tests ... done in 6 minutes and 35 seconds
(5/13) Running minio-go tests ... FAILED in 9 minutes and 57 seconds
{
  "args": {
    "bucketName": "minio-go-test-4lokx9t6iqzsw41o",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 218,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

File: cmd/xl-sets.go [hunk 1]
```diff
@@ -1045,9 +1045,10 @@ func (s *xlSets) listObjectsNonSlash(ctx context.Context, bucket, prefix, marker
 	return result, nil
 }
 
-// ListObjects - implements listing of objects across disks, each disk is indepenently
+// ListObjects - implements listing of objects across disks, each disk is independently
 // walked and merged at this layer. Resulting value through the merge process sends
 // the data in lexically sorted order.
+// If heal is set only objects that does not have full quorum is returned.
```
Comment (vadmeste) [inline cmd/xl-sets.go]: Could you rename 'heal' to 'avoidFullQuorum' here and in xlSets.listObjects() as well ?

**Review by vadmeste — COMMENTED** [review]
LGTM but with one simple comment

File: cmd/xl-sets.go (hunk 1 above)
Comment (klauspost) [reply cmd/xl-sets.go]: Good idea.

Comment (klauspost) [reply cmd/xl-sets.go]: @vadmeste Went for `partialQuorumOnly`.

Also the field was unused in listParams, so I removed it.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#592312ff-ed63-4384-9288-049369b29d72) |
| mint-dist-xl.sh | :x: [more...](#bb70b760-f76c-4bc2-b3a4-fbe497a5c613) |
| mint-gateway-azure.sh | :x: [more...](#a635aa8a-2cfd-41a8-bed3-dc1392f47442) |
<a name="bb70b760-f76c-4bc2-b3a4-fbe497a5c613"></a>
### 8900-2467e48/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30522
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp daaabe4cc850:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 8 seconds
{"name": "awscli", "duration": 305395, "function": "make_bucket\n", "status": "FAIL", "error": "Connect timeout on endpoint URL: "http://72.28.97.55:30522/awscli-mint-test-bucket-12854""}

Executed 0 out of 13 tests successfully.
```
<a name="a635aa8a-2cfd-41a8-bed3-dc1392f47442"></a>
### 8900-2467e48/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:32091
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3295ee91ee6c:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 14 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 16 seconds
(4/13) Running mc tests ... done in 6 minutes and 11 seconds
(5/13) Running minio-go tests ... FAILED in 20 minutes and 51 seconds
{
  "args": {
    "bucketName": "minio-go-test-1kxofolahs01zj3b",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 403,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="592312ff-ed63-4384-9288-049369b29d72"></a>
### 8900-2467e48/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32575
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d7d119843a4e:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 29 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 7 seconds
(4/13) Running mc tests ... done in 49 seconds
(5/13) Running minio-go tests ... FAILED in 4 minutes and 24 seconds
{
  "args": {},
  "duration": 138759,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#d6369ba7-19ff-46eb-8246-b9a2b0b27ef2) |
| mint-large-bucket.sh | :x: [more...](#57ed5e69-68b5-414f-9a01-d787f987819c) |
| mint-gateway-azure.sh | :x: [more...](#b57c5b0f-63f4-4c64-8fc9-a297d6532c55) |
<a name="d6369ba7-19ff-46eb-8246-b9a2b0b27ef2"></a>
### 8900-f0d8e53/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31714
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 388a70e5f5bd:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 11 seconds
{
  "name": "awscli",
  "duration": 8899,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="b57c5b0f-63f4-4c64-8fc9-a297d6532c55"></a>
### 8900-f0d8e53/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32173
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6620e229d89f:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 21 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 17 seconds
(4/13) Running mc tests ... done in 7 minutes and 14 seconds
(5/13) Running minio-go tests ... FAILED in 10 minutes and 30 seconds
{
  "args": {
    "bucketName": "minio-go-test-39u1pbmctgjnw2fl",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 232,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="57ed5e69-68b5-414f-9a01-d787f987819c"></a>
### 8900-f0d8e53/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:30577
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9e5bcdcd7a08:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 33 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 1 minutes and 2 seconds
(5/13) Running minio-go tests ... FAILED in 4 minutes and 54 seconds
{
  "args": {},
  "duration": 130927,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by klauspost** [comment]
@vadmeste  Look, now @harshavardhana Got #8901 in first, so I have to deal with the merge conflicts. :/

**Review by harshavardhana — APPROVED** [review]

**Comment by klauspost** [comment]
This just ended up as a cleanup job, no functional changes after #8901

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#903da524-0715-4cb7-9ee4-6d162ed33a28) |
| mint-gateway-azure.sh | :x: [more...](#409b37fc-f79a-452f-9f11-7a4011870426) |
<a name="409b37fc-f79a-452f-9f11-7a4011870426"></a>
### 8900-a4cb96b/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:30563
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 996ba3f4f241:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 7 seconds
{"name": "awscli", "duration": 304415, "function": "make_bucket\n", "status": "FAIL", "error": "Connect timeout on endpoint URL: "http://72.28.97.55:30563/awscli-mint-test-bucket-31410""}

Executed 0 out of 13 tests successfully.
```
<a name="903da524-0715-4cb7-9ee4-6d162ed33a28"></a>
### 8900-a4cb96b/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.58:31193
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1d792cbb3770:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 32 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 44 seconds
(5/13) Running minio-go tests ... FAILED in 13 minutes and 54 seconds
{
  "args": {},
  "duration": 104631,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#4a86f0b7-b829-403d-bc98-28d288eb1568) |
| mint-dist-xl.sh | :x: [more...](#a6368664-4f30-49a1-a814-048fd42c31cd) |
| mint-gateway-azure.sh | :x: [more...](#96258fa6-74f0-4a72-8091-6c1827786455) |
<a name="4a86f0b7-b829-403d-bc98-28d288eb1568"></a>
### 8900-9c93856/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:30237
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e65097a1a1d5:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 23 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 43 seconds
(5/13) Running minio-go tests ... FAILED in 13 minutes and 39 seconds
{
  "args": {},
  "duration": 128982,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="a6368664-4f30-49a1-a814-048fd42c31cd"></a>
### 8900-9c93856/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31690
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp d034c1981258:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 12 seconds
{"name": "awscli", "duration": 309637, "function": "make_bucket\n", "status": "FAIL", "error": "Connect timeout on endpoint URL: "http://72.28.97.55:31690/awscli-mint-test-bucket-29450""}

Executed 0 out of 13 tests successfully.
```
<a name="96258fa6-74f0-4a72-8091-6c1827786455"></a>
### 8900-9c93856/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:30285
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ef580ca8c956:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 3 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 17 seconds
(4/13) Running mc tests ... done in 6 minutes and 46 seconds
(5/13) Running minio-go tests ... FAILED in 22 minutes and 3 seconds
{
  "args": {
    "bucketName": "minio-go-test-ti503a6z9q3cbxoj",
    "objectPrefix": "",
    "recursive": "true"
  },
  "duration": 238,
  "error": "The specified bucket is not valid.",
  "function": "ListObjects(bucketName, objectPrefix, recursive, doneCh)",
  "message": "PutObject 1 call failed",
  "name": "minio-go: testListObjects",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by harshavardhana** [comment]
PTAL @vadmeste

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**